source/js/searcher.js

```diff
@@ -0,0 +1,771 @@
+//  MIT License
```

Comment:
This whole file is impressive! My review comments below will be assuming that future students will be maintaining the code, so it needs to be extra simple to grasp (so lots of extra explanation and extra clear variable naming should be in place).

Reply:
done